text/0000-third-party-web-components-support.md

```diff
@@ -0,0 +1,103 @@
+be delegated to the browser. LWC will not attempt to resolve components for custom elements tagged
+with this directive. If it happens that the component was not registered, the custom element will
+simply render as an HTMLUnknownElement, which is the default browser behavior.
+
+### Passing data
+
+By convention, custom elements are assumed to be backed by LWC components and values are set through
+their properties. Given that a native web component might expect data passed as an attribute or as a
+property, we would need to support both mechanisms. There are multiple approaches we could take
+here, and the following options are how some popular frameworks do it today. We could default to
+either attributes or properties and introduce a directive that would act as a signal to the engine
+to use one or the other, we could implement a runtime heuristic to choose one over the other, or we
+could default to attributes and set properties when setting a data structure.
+
+#### Proposed solution
+
+When passing data to a native web component, we can take a heuristical approach where if the
+property exists, the property is set. Otherwise the attribute is set. While it may be advantageous
+to require an explicit signal in the form of a directive, this approach seems to strike a good
+balance between developer experience and API semantics.
+
+### Events
+
+EventTarget polyfilled methods and properties such as `add/removeEventListener` and `target` go
+through the synthetic shadow polyfill due to those methods and properties being globally patched.
+This is the current status quo for native shadow roots and should not be an issue for supporting
+third party web components.
+
+LWC's declarative event bindings only support lowercase events. Events using anything other than
+lowercase must be listened to imperatively. This would not need to change in order to support native
+web components.
+
+### Serverside rendering (SSR)
+
+LWC's SSR implementation is tailored specifically for LWC and may not be compatible with a native
+web component implementation. We would probably need to implement the entire synchronous DOM API if
+we wanted to support SSR for native web components.
+
+## Drawbacks
+
+As with any new feature, adding support for native web components introduces additional complexity.
+
+From the platform perspective, there are a few things that Locker vNext needs to implement in order
+to GA this feature. 1) Lift the restriction on `customElements.define()` and 2) provide the
+virtualization mechanism to allow different namespaces to use the same component name.
+
+## Alternatives
+
+No alternatives were considered for the constructor resolution portion of this proposal. An initial
+uninformed idea to introduce error handling for implicit constructor imports was quickly eliminated
+due to the fact that dynamic imports are promise-based.
+
+Alternatives for passing data have already been discussed.

```

Comment:
Did we consider the alternative of not requiring an additional directive, but having some kind of fallback logic? E.g.:

1. The developer uses `<x-foo></x-foo>` in a template
2. First, we check if `x-foo` is an LWC component
3. If not, we check if `x-foo` is defined in the global registry
4. If not, we throw an error

Is something like this feasible?

Reply:
> * The developer uses `<x-foo></x-foo>` in a template

If the template contains an `<x-foo>` element, the LWC compiler would automatically add an import for the `x/foo` module. If the `x/foo` module doesn't exist the compilation would fail. While you can always create this module in LWC OSS, platform developers only have access to their own `c` namespace.

We need a mechanism to tell the compiler not to attempt to import certain custom elements to avoid running into those resolution issues.